cdap-unit-test/src/test/java/co/cask/cdap/security/AuthorizationTest.java

```diff
@@ -839,6 +843,131 @@ public Boolean call() throws Exception {
+      flowManager.start();
+      Assert.fail("Bob don't have the privlege to write to the dataset. Failed");
+    } catch (Exception e) {
+      // Expected
+    }
+
+    // Switch back to ALICE. Should Success
+    SecurityRequestContext.setUserId(ALICE.getName());
+    FlowManager flowManager = appManager.getFlowManager(CrossNsDatasetAccessApp.FLOW_NAME);
+    flowManager.start();
+    for (int i = 0; i < 10; i++) {
+      streamManager.send(String.valueOf(i).getBytes());
+    }
+    flowManager.getFlowletMetrics("saver").waitForProcessed(10, 30, TimeUnit.SECONDS);
+
+    DataSetManager<KeyValueTable> dataSetManager = getDataset(datasetOutputSpace.toId(), "store");
+    KeyValueTable results = dataSetManager.get();
+    for (int i = 0; i < 10; i++) {
+      byte[] key = String.valueOf(i).getBytes(Charsets.UTF_8);
+      Assert.assertArrayEquals(key, results.read(key));
+    }
+
+    flowManager.stop();
+    getNamespaceAdmin().delete(datasetOutputSpace.toId());
+  }
+
+  @Test
+  public void testCrossNSDatasetAccessWithAuthMapReduceSpark() throws Exception {
+    createAuthNamespace();
+    getNamespaceAdmin().create(new NamespaceMeta.Builder()
+                                 .setName(DatasetCrossNSAccessWithMAPApp.DATASET_INPUT_SPACE).build());
+    getNamespaceAdmin().create(new NamespaceMeta.Builder()
+                                 .setName(DatasetCrossNSAccessWithMAPApp.DATASET_OUTPUT_SPACE).build());
+    NamespaceId datasetInputSpace = new NamespaceId(DatasetCrossNSAccessWithMAPApp.DATASET_INPUT_SPACE);
+    NamespaceId datasetOutputSpace = new NamespaceId(DatasetCrossNSAccessWithMAPApp.DATASET_OUTPUT_SPACE);
+
+    addDatasetInstance(datasetInputSpace.toId(), "keyValueTable", "table1").create();
+    addDatasetInstance(datasetOutputSpace.toId(), "keyValueTable", "table2").create();
+    DataSetManager<KeyValueTable> tableManager = getDataset(datasetInputSpace.toId(), "table1");
+    KeyValueTable inputTable = tableManager.get();
+    inputTable.write("hello", "world");
+    tableManager.flush();
+
+    ApplicationManager appManager = deployApplication(AUTH_NAMESPACE.toId(), DatasetCrossNSAccessWithMAPApp.class);
+    Map<String, String> argsForMR = ImmutableMap.of(DatasetCrossNSAccessWithMAPApp.INPUT_KEY, "table1",
+                                                    DatasetCrossNSAccessWithMAPApp.OUTPUT_KEY, "table2");
+    MapReduceManager mrManager = appManager.getMapReduceManager(DatasetCrossNSAccessWithMAPApp.MAPREDUCE_PROGRAM);
+    // Switch to Bob. Should fail as Bob do not have access to the dataset namespace
+    getAuthorizer().grant(AUTH_NAMESPACE, BOB, ImmutableSet.of(Action.ALL));
+    SecurityRequestContext.setUserId(BOB.getName());
+    try {
+      mrManager.start(argsForMR);
+      mrManager.waitForFinish(5, TimeUnit.MINUTES);
+      Assert.fail("Should fail here since bob does not have the privileges on input dataspace.");
+    } catch (Exception e) {
+      // Expected
+    }
+
```

Comment:
can we make sure that the program actually failed? e.g. that the output dataset did not get written to?